### PR TITLE
gh-91603: Speed up UnionType instantiation (alt)

### LINKED ISSUE #91603: Speed up isinstance on union types
[PEP 604](https://peps.python.org/pep-0604/#isinstance-and-issubclass) allows `isinstance` on union types but it much slower compared to a tuple of types. This can be speed up with a fast path. Currently it is almost 4x slower, so there is room for improvement.

Benchmarks:
```console
@kumaraditya303 ➜ /workspaces/cpython (fast-iter-str) $ ./python -m timeit "isinstance(1, (int, str))"
5000000 loops, best of 5: 54.6 nsec per loop
@kumaraditya303 ➜ /workspaces/cpython (fast-iter-str) $ ./python -m timeit "isinstance(1, int|str)"
1000000 loops, best of 5: 202 nsec per loop
```

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Co-authored-by: Yurii Karabas <1998uriyyo@gmail.com>

It is a rewriting of #91865.

Closes #91603.

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
The new code is even shorter than the old one!
```
 1 file changed, 77 insertions(+), 87 deletions(-)
```

**Comment by uriyyo** [comment]
@serhiy-storchaka Just compare performance with my version. It's almost identical:
| Benchmark                                                                                                                                          | yurii    | serhiy                |
|----------------------------------------------------------------------------------------------------------------------------------------------------|:-------:|:---------------------:|
| int \| float                                                                                                                                        | 68.8 ns | 73.2 ns: 1.06x slower |
| int \| (list \| float)                                                                                                                               | 136 ns  | 139 ns: 1.02x slower  |
| float \| set \| int \| int \| float \| list                                                                                                             | 307 ns  | 288 ns: 1.06x faster  |
| bool \| int \| float \| complex \| str \| bytes \| bytearray \| list \| set \| frozenset \| dict \| range \| property \| classmethod \| staticmethod \| Exception | 1.41 us | 1.61 us: 1.14x slower |
| Geometric mean                                                                                                                                     | (ref)   | 1.02x slower          |

Benchmark hidden because not significant (3): (int | float) | list, (int | float) | (list | set), (float | set) | int | int | (float | list)

File: Objects/unionobject.c [hunk 1]
```diff
@@ -242,9 +229,23 @@ _Py_union_type_or(PyObject* self, PyObject* other)
         Py_RETURN_NOTIMPLEMENTED;
     }
 
-    PyObject *tuple = PyTuple_Pack(2, self, other);
+//     if (self == Py_None) {
+//         self = (PyObject *)&_PyNone_Type;
+//     }
+//     if (other == Py_None) {
+//         other = (PyObject *)&_PyNone_Type;
+//     }
```
Comment (uriyyo) [inline Objects/unionobject.c]: Just a reminder to remove this code. It will make code even more shorter than it was)

**Review by uriyyo — APPROVED** [review]
LGTM

**Comment by serhiy-storchaka** [comment]
I am not surprised because it is the same algorithm.

"type | type" may be slightly slower because the code is more general, but it is a honest price for simpler code. On other hand, my code avoid creation a new tuple and a new UnionType object if possible, so it is expected that cases like `float | set | int | int | float | list` may be slightly faster.

I was skeptical about advisability of such optimization at all, but if it also reduces the size of the code I have no doubts.

File: Objects/unionobject.c [hunk 2]
```diff
@@ -137,93 +137,80 @@ union_richcompare(PyObject *a, PyObject *b, int op)
...
-static PyObject*
-dedup_and_flatten_args(PyObject* args)
+static PyObject **
+get_types(PyObject **obj, Py_ssize_t *size)
 {
-    args = flatten_args(args);
-    if (args == NULL) {
-        return NULL;
+    if (*obj == Py_None) {
+        *obj = (PyObject *)&_PyNone_Type;
```
Comment (JelleZijlstra) [inline Objects/unionobject.c:203]: Don't we need to INCREF here?

**Review by JelleZijlstra — COMMENTED** [review]
Looks mostly good, but I have a question about refcounts.

File: Objects/unionobject.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Objects/unionobject.c:203]: No, it is a borrowed reference.

**Pull request merged**